### ISSUE TITLE
Naming conventions of surface templates

### ISSUE THREAD

**Issue by feilong** [templateflow/templateflow#171]
When we say surface templates/spaces, we can mean two different things.
1. Registration target: A standard sphere and the alignment data associated with the sphere (e.g., sulc, curv).
2. Resampling target: A set of standard vertices on the standard sphere.

The registration target and resampling target are often the same for volumetric data, but they can be different for surface data.  For example, after obtaining `sphere.reg` with FreeSurfer (i.e., using FreeSurfer's registration target), we can use it to resample data to different resampling targets (e.g., `fsaverage`, `fsaverage5`).

To me it's confusing when we use the same name for both registration target and resampling target, e.g., using `fsaverage` to refer to both FreeSurfer's registration target and its resampling target with icoorder7.

I think it's better to name the registration targets based on the group of participants used to generate the alignment data (e.g., `fsavg40`, `conte69`, `hcp196`), and name the resampling targets with density/resolution (e.g., `fslr-32k`, `fsavg-ico32`).  Does it make sense to people?

Some thoughts related to our discussion today @oesteban @effigies.  I guess what Oscar wants is transformation between registration targets, and what I want is transformation between resampling targets.

**Comment by tsalo** [comment]
@feilong I think this is a natural place to follow up on https://github.com/nipreps/fmriprep/issues/3330#issuecomment-2247075427.

Just so I understand, there seem to be three issues here:

1. TemplateFlow's file naming doesn't distinguish between registration targets and resampling targets.
    - Does the `desc-reg` in some sphere filenames mean "registration target"?
2. Registration target filenames should include the sample name.
    - Do registration targets from different samples still correspond to the same template? If so, then I think an entity would be the way to go, rather than a distinct template name.
3. Resampling targets should include the density/resolution.
    - Don't they already? E.g., `tpl-fsLR_hemi-L_den-164k_sphere.surf.gii`

**Comment by feilong** [comment]
Yes, I think it makes more sense to continue the discussion here. Here is my understanding:

1. `desc-reg` means the sphere is registered to the registration target, which differentiate it from the original sphere. In FreeSurfer terms, it's the difference between `sphere.reg` and `sphere`.
2. Some registration targets correspond to the same template, usually fsLR. However, because they were derived from different samples, the folding patterns slightly differ, which means using different registration targets would lead to different results.
3. Yes, density/resolution is already in the filename. In the post above, I just wanted to mention that density is only relevant to the resampling target, not the registration target.

**Comment by tsalo** [comment]
1. Okay that makes sense. Does the `space` entity indicate the same thing? E.g., `tpl-fsLR_space-fsaverage_sphere.surf.gii`? If so, is that sufficient to distinguish registration from resampling spheres?
2. I guess the question becomes whether the different flavors of registration target are different enough to be considered different templates (like the many kinds of MNI152 template) or if an entity would be sufficient (e.g., `tpl-fsaverage_sample-conte69_sphere.surf.gii`). WDYT?

**Comment by feilong** [comment]
1. I think so. In this example `tpl-fsLR` is the resampling target, and `space-fsaverage` is the registration target.
2. I'm inclined to treat them as different templates, mainly because the registration results would also differ for these registration targets based on different groups.

**Comment by Shotgunosine** [comment]
I'm not sure I entirely follow on point 1. @tsalo, would you mind spelling out exactly what you're proposing? Actually, it'd be easier for me to comment if you could sum up what the proposed naming scheme would be. I just want to avoid making assumptions and inadvertently creating a new set of possibilities.

**Comment by tsalo** [comment]
Sorry, I think I was a little mixed up. My current understanding is that templateflow contains (or should contain) three types of sphere:

1. Transform/registration sphere, where the vertices correspond to the source sphere, but the coordinates reflect the position on the target sphere. For example, `tpl-fsLR_space-fsaverage_hemi-L_den-32k_sphere.surf.gii`.
    - I think I was mixing this up with a "registration target".
2. Resampling target sphere, where the vertices, mesh, and coordinates describe the template at a given density. For example, `tpl-fsLR_hemi-L_den-32k_sphere.surf.gii`.
3. Registration target sphere, which I'm still a little confused about.
    - Multiple registration targets can be created for the same general space (e.g., fsLR), but from different samples (e.g., `hcp449`), which can produce different folding patterns. Therefore, @feilong is proposing that these spheres be differentiated based on which sample was used to create each.
    - @feilong are there any examples of registration target spheres on templateflow already?

The main issue is that resampling targets and registration targets are not differentiated in templateflow.

**Comment by feilong** [comment]
I agree with the idea of three kinds of spheres. Here is my understanding:

1. Registration target. This is often a combination of one or more morphometric measures (e.g., sulc, curv) and a sphere. The measures and the sphere have correspondence, which means they have the same number of vertices.
2. Resampling target. It is always relative to a certain registration target. In other words, its vertices are defined on the same sphere of the registration target. I think `tpl-fsLR_space-fsaverage_hemi-L_den-32k_sphere.surf.gii` is actually a resampling target, and the vertices (of the fsLR template) are defined on the fsaverage sphere.
3. Transform. It should be a pair of spheres with the same number of vertices. The transform is defined based on the locations of the same set of vertices on different spheres. Pairs of resampling targets can be used as transforms. In practice, it's better to use high-density spheres as transforms. For example, HCPpipelines use 164k to transform between fsaverage and fsLR.

**Comment by tsalo** [comment]
Wow, I was _way_ off! Apologies everyone. @Shotgunosine, does @feilong's answer match your understanding?

**Comment by Shotgunosine** [comment]
Just to be clear, @feilong is the biggest reason I've got any understanding of this stuff. I think what he's said is correct, but maybe I can expand a little bit.

# File types
The HCP naming convention for gifti surfaces includes `.surf.gii` and `.shape.gii` files. 
`.surf.gii` files define a mesh and two arrays, one defining the x,y,z coordinates of each vertex, and one listing all of the triplets of vertex indices that define the triangles making up the mesh. All of the surfaces (pial, white matter, inlated, spherical) for one hemisphere from a subject will have the same number of vertices, and the same array of triangles, but different coordinates for each vertex. 
`.shape.gii` files have a single array of values, one for every vertex. So you can have thickness, curvature, myelin maps, connectivity maps, whatever, each in a `.shape.gii` file.

# Registration
The goal of spherical registration like `MSMAll` is to deform one sphere so that it matches another sphere, usually you'd be deforming a subject's sphere so that it matches the sphere from some template space. MSMAll can simultaneously take a bunch of different metrics into account when aligning two spheres, assuming you've got `.shape.gii` files defining that metric for both the subject and the template sphere. In this case the "registration target" is the pattern of all of the metrics you're using on the template sphere. The template `.surf.gii` file defines the spherical mesh, and the template's `.shape.gii` files define the values of those metrics on the template sphere, so the registration target is defined by a collection of files. The output of registration will be a sphere with the same number of vertices as the subject's mesh, with coordinates that minimize the difference between the subject's metrics and the template metrics, this is the `_desc-reg_sphere.surf.gii`.

## Transformation
The transformation itself is the difference between the coordinate arrays in the subject's `_sphere.surf.gii` and the `_desc-reg_sphere.surf.gii`. The transforms that @feilong is talking about are the transformations between template spaces, but I'm not going to get into that right now. 

# Resampling
Great, now we've got a registration, but what if we want to compare cortical thickness values across individuals. Our subject's cortical thickness `.shape.gii` file just has one value per vertex, and the meshes for each subject are different. The `_desc-reg_sphere.surf.gii` has the coordinates for each of those vertices in the template space, but in order to compare cortical thicknesses, we need to get all of our subjects cortical thickness values on the same mesh. For this, we need a sphere in the template space with the mesh we want to use for all of our subjects. If we're in `space-fsLR`, we can pick between 3 different meshes with different densities of vertices. If we want a density that's roughly equivalent to a 2mm voxel size, we can pick `den-32k`, so the file we need is `tpl-fsLR_hemi-R_den-32k_sphere.surf.gii`. The resampling works by finding which triangle in the subject's mesh encloses each vertex in the template mesh. The cortical thickness value at each vertex in the template mesh is interpolated from the cortical thickness values at the 3 vertices from the subject mesh that make up the enclosing triangle.

@tsalo, does that help at all? If so, what are the outstanding questions regarding naming surface templates? @feilong, would you mind restating what your renaming suggestion is?

**Comment by feilong** [comment]
Thanks for the super clear explanation, @Shotgunosine.

My suggestion is about distinguishing different registration targets, mainly different flavors of fsLR. Initially fsLR was created by aligning the left and right hemispheres of fsaverage, so that the vertices of the two hemispheres have correspondence. In other words, this definition is a transformation between fsaverage and fsLR. After that, people have created a few different registration targets for fsLR, based on different groups of participants. These registration targets slightly differ, because the averaged morphometric features (`.shape.gii` files) of them were based on different groups of participants. For example, `conte69`, `hcp196`, and `hcp449`.

The `_desc-reg_sphere.surf.gii` (or `sphere.reg`) file of a participant is determined by both the registration target (e.g., `conte69`, `hcp196`, `hcp449`) and the registration algorithm (e.g., FreeSurfer, MSMSulc, MSMAll). In other words, if we use MSM with different registration targets, the resulting sphere would be different. In my opinion, it is better to distinguish them in the file name, given the file content would be different.

Regarding the registration target, my suggestion is to give each registration target a unique name, such as a combination of the name of the dataset and the number of individuals.

To distinguish different `_desc-reg_sphere.surf.gii` files, we might also want to include the name of the registration algorithm in the filename.

**Comment by NingAnMe** [comment]
> Thanks for the super clear explanation, @Shotgunosine.
> 
> My suggestion is about distinguishing different registration targets, mainly different flavors of fsLR. Initially fsLR was created by aligning the left and right hemispheres of fsaverage, so that the vertices of the two hemispheres have correspondence. In other words, this definition is a transformation between fsaverage and fsLR. After that, people have created a few different registration targets for fsLR, based on different groups of participants. These registration targets slightly differ, because the averaged morphometric features (`.shape.gii` files) of them were based on different groups of participants. For example, `conte69`, `hcp196`, and `hcp449`.
> 
> The `_desc-reg_sphere.surf.gii` (or `sphere.reg`) file of a participant is determined by both the registration target (e.g., `conte69`, `hcp196`, `hcp449`) and the registration algorithm (e.g., FreeSurfer, MSMSulc, MSMAll). In other words, if we use MSM with different registration targets, the resulting sphere would be different. In my opinion, it is better to distinguish them in the file name, given the file content would be different.
> 
> Regarding the registration target, my suggestion is to give each registration target a unique name, such as a combination of the name of the dataset and the number of individuals.
> 
> To distinguish different `_desc-reg_sphere.surf.gii` files, we might also want to include the name of the registration algorithm in the filename.

Hi, @feilong,
Are the registration targets differentiated here, similar to `tpl-MNI152NLin2009aAsym`, `tpl-MNI152NLin2009aSym`, `tpl-MNI152NLin6Asym`, `tpl-MNI152Lin`? For example, `tpl-fsLRconte69`, `tpl-fsLRhcp196`, `tpl-fsLRhcp449`?

另外I have a question and hope to receive your reply. What is the overall meaning of the naming convention `tpl-fsLR_space-fsaverage_hemi-L_den-164k_sphere.surf.gii`? Does it mean that the `tpl-fsLR` template has been sampled to `tpl-fsaverage_hemi-L-den-164k_sphere`? If so, how are `tpl-fsLR_space-fsaverage_hemi-L_den-32k_sphere.surf.gii` and `tpl-fsLR_space-fsaverage_hemi-L_den-59k_sphere.surf.gii` derived? The `tpl-fsaverage `template does not have `den-32k` or `den-59k`.

**Comment by feilong** [comment]
Hi @NingAnMe, the examples here (`conte69`, `hcp196`, and `hcp449`), are more like the differences between `MNI152NLin2009aAsym` and `MNI152NLin2009cAsym`.  They share the same objective (somewhat like `Asym` and `NLin` for MNI) for the optimization process, but the results differ, mainly because they were based on different groups of participants.

I believe `tpl-fsLR` and `den-32k` defines the template and the resolution. `space-fsaverage` means the sphere has been warped to match the `fsaverage` sphere.  That's similar to the difference between `sphere` and `sphere.reg`.

**Comment by Shotgunosine** [comment]
I think naming-wise, following the example of MNI152 makes sense. As far as I know though, BIDS doesn't usually keep track of which program's nonlinear registration was used in the file name, would that make more sense as a json field?

**Comment by feilong** [comment]
I was thinking about having the algorithm/program as the `desc-` part of the filename.  A json field also makes sense to me.
